### ISSUE TITLE
Issues Exporting Voice Data from FACEIT Demos After Recent Major CS2 Update (2/6/2024)

### ISSUE THREAD

**Issue by justMemou** [DandrewsDev/CS2VoiceData#4]
Hello,

Since the latest major update, I've been encountering issues while exporting voice data from FACEIT demos. During the analysis, I am overwhelmed with notifications indicating invalid voice packets. Here is an example of the notifications I receive:
```
invalid voice packet (received 1a, expected b)
invalid voice packet (received ec, expected b)
invalid voice packet (received 58, expected b)
invalid voice packet (received e4, expected b)
invalid voice packet (received 7b, expected b)
invalid voice packet (received 65, expected b)
invalid voice packet (received 50, expected b)
invalid voice packet (received 37, expected b)
invalid voice packet (received 11, expected b)
invalid voice packet (received 33, expected b)
invalid voice packet (received c2, expected b)
invalid voice packet (received 4d, expected b)
invalid voice packet (received fd, expected b)
invalid voice packet (received fb, expected b)
invalid voice packet (received f3, expected b)
invalid voice packet (received f7, expected b)
invalid voice packet (received f3, expected b)
invalid voice packet (received e1, expected b)
invalid voice packet (received 28, expected b)
invalid voice packet (received 22, expected b)
invalid voice packet (received 87, expected b)
invalid voice packet (received b1, expected b)
invalid voice packet (received ba, expected b)
invalid voice packet (received f7, expected b)
invalid voice packet (received 49, expected b)
invalid voice packet (received d6, expected b)
invalid voice packet (received 8c, expected b)
invalid voice packet (received f2, expected b)
invalid voice packet (received 16, expected b)
invalid voice packet (received f7, expected b)
invalid voice packet (received c3, expected b)
invalid voice packet (received d9, expected b)
invalid voice packet (received ed, expected b)
invalid voice packet (received b9, expected b)
invalid voice packet (received 25, expected b)
```
Despite these errors, WAV files are created at the end of the analysis process, but they are blank.

I am curious if anyone else has encountered this issue. If so, is there a plan to update the library to address this problem?

Thank you.

**Comment by DandrewsDev** [comment]
Hi, I've confirmed this issue is occurring with demos created after the "A Call to Arms" update. 
I am currently looking into a fix.

**Comment by DandrewsDev** [comment]
The data we're getting back from the demo is in a "new" format. Previously it was a steam audio format which contained a few chunks of info like the users steam id, payload type, size, and checksum. Which was then followed by opus data.

The new format is just called opus and seemingly ditches the additional information. I've tried adjusting the chunking and decoder to just look for the opus info, but no luck as of yet.

Old
audio:{format:VOICEDATA_FORMAT_STEAM  voice_data:"059c8307010010010bc05d060a0155007500689beb6eb96bb4ffff5fb28febdd8c607cb643485b85574bfc7985b291a8983965c56ffc8a67153c315c65cc269547c2070789c2ad795da022a6bacb365a1f8e459993ab7b4c986d74a8295a3c408531528761303558007600689b58e2fce89e89b76f02e6c0228b48ee6b6ced22e96759a3fb3597d3bcc5b107ab114f053c32017763d275ffb3d07298aed8ea3c04e297d20ebd631759543d734d45b95693a5a3e30f72b8ec6d2ff6b511e0ff807dfb165100770068a302ab551306ef750c8ab7fbaf4bc9999cd726b8f7c2dd69c5c38c22ded6694f4388121659d0407796d26f4cd5cc85a4bf769bbf8da25a8dcc478043ab0a73415f6caf26eefb7cc7a0746f4c36d83205072dc11b"  sequence_bytes:10886  section_number:298  sample_rate:24000}  client:7  proximity:false  xuid:76561198086331397  audible_mask:1 

New
audio:{format:VOICEDATA_FORMAT_OPUS  voice_data:"70041ed99c2af8cce8d1defd5676b8738270627290c158645a7f53284f266156eeb8a3f91a2f076008c5c41452ccdc766ca6e7fce71119ac7fa931b0c528ea958213e338670ed5"  sequence_bytes:0  section_number:793  sample_rate:48000  num_packets:1  packet_offsets:71  packet_offsets:0  packet_offsets:0  packet_offsets:0  voice_level:-27.818909}  client:12  proximity:false  xuid:76561199203408612  audible_mask:1 

I'm continuing to work the issue, but am open to any PRs, input, or info anyone might have.

**Comment by justMemou** [comment]
Hmm *just thinking out loud* I remember previously (in the CSGO days) the data was decoded with the game .so files with
ref: `https://gist.github.com/ericek111/abe5829f6e52e4b25b3b97a0efd0b22b`
My previous demo analyser never made any checks, just dump the data and into a string parse it with the game libraries. Do you think this might be the case here?

**Comment by maxpain** [comment]
@justMemou any updates?

**Comment by DandrewsDev** [comment]
No updates as of yet. I've tried a ton of different changes to how data was being chunked and processed to account for the new format. I've yet to be able to decode the new opus data into any meaningful audio output.

I also considered using the .so files included with the game. Though those aren't as simple as just getting a string and passing it along.

I reached out to valve for any info on the updated format, but no response as of yet. 

This is going to take some real work to fix, I've only been able to spend a little time on weekends working on it. But hopefully we'll have a fix soon.

**Comment by maxpain** [comment]
```go
package main

import (
	"fmt"
	"log"
	"os"

	. "voice/decoder"

	"github.com/go-audio/audio"
	"github.com/go-audio/wav"
	dem "github.com/markus-wa/demoinfocs-golang/v4/pkg/demoinfocs"
	"github.com/markus-wa/demoinfocs-golang/v4/pkg/demoinfocs/msgs2"
)

func main() {
	file, err := os.Open("12388254_12463469_2402192025-de_inferno.dem")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	parser := dem.NewParser(file)
	defer parser.Cancel()

	voiceDataPlayers := make(map[uint64][][]byte, 0)

	parser.RegisterNetMessageHandler(func(packetData *msgs2.CSVCMsg_VoiceData) {
		voiceDataPlayers[*packetData.Xuid] = append(voiceDataPlayers[*packetData.Xuid], packetData.Audio.VoiceData)
	})

	err = parser.ParseToEnd()
	if err != nil {
		panic(err)
	}
	defer parser.Close()

	for steamID64, voiceData := range voiceDataPlayers {
		wavName := fmt.Sprintf("%d.wav", steamID64)

		err = opusToWav(voiceData, wavName)
		if err != nil {
			log.Println(err)
			continue
		}
	}
}

func opusToWav(data [][]byte, wavName string) (err error) {
	decoder, err := NewDecoder(48000, 1)
	if err != nil {
		return
	}

	var pcmBuffer []int

	for _, d := range data {
		pcm, err := Decode(decoder, d)
		if err != nil {
			log.Println(err)
			continue
		}

		pp := make([]int, len(pcm))

		for i, p := range pcm {
			pp[i] = int(p * 2147483647)
		}

		pcmBuffer = append(pcmBuffer, pp...)
	}

	file, err := os.Create(wavName)
	if err != nil {
		return
	}
	defer file.Close()

	enc := wav.NewEncoder(file, 48000, 32, 1, 1)
	defer enc.Close()

	buffer := &audio.IntBuffer{
		Data: pcmBuffer,
		Format: &audio.Format{
			SampleRate:  48000,
			NumChannels: 1,
		},
	}

	err = enc.Write(buffer)
	if err != nil {
		return
	}

	return
}

```

`decoder.go`:
```go
package decoder

import (
	"gopkg.in/hraban/opus.v2"
)

func NewDecoder(sampleRate, channels int) (decoder *opus.Decoder, err error) {
	decoder, err = opus.NewDecoder(sampleRate, channels)
	return
}

func Decode(decoder *opus.Decoder, data []byte) (pcm []float32, err error) {
	pcm = make([]float32, 1024)

	nlen, err := decoder.DecodeFloat32(data, pcm)
	if err != nil {
		return
	}

	return pcm[:nlen], nil
}
```

**Comment by DandrewsDev** [comment]
@maxpain I can confirm that fix works for the newest demos. I appreciate this a lot, if you want to make a PR feel free and I'll accept those changes. Or I just just pull in your code and make those changes if you prefer.

**Comment by maxpain** [comment]
@DandrewsDev you can do it.

Seems like voice in CS2 demos now has sample rate 48000 Hz, before update it was 24000 Hz. Isn't it?

**Comment by DandrewsDev** [comment]
@maxpain Correct. Im going to make a branch and a PR here in a few with code that will auto detect which decoding method it use use by checking the format variable from the voice data event. 

That was the code sample will work with any of the CS2 demos both before and after this change.

**Comment by DandrewsDev** [comment]
Closing as fixed with PR #5
